File: JavaScript/test/recursive_permutations.js
const chai = require('chai');
const expect = chai.expect;

const getPermutations = (str) => {
  if(str.length <= 1) return new Set([str]);

  let results = new Set();
  str.split('').forEach((char, i) => {
    let remaining = str.slice(0, i) + str.slice(i + 1);
    let perms = getPermutations(remaining);
    perms.forEach(p => {
      results.add(char + p);
    });
  });
  return results;
};

describe('Recursive String Permutations', () => {
  it('empty string', () => {
    let input = '';
    let actual = getPermutations(input);
    let expected = new Set(['']);
    expect(actual).to.deep.equal(expected);
  });

  it('one character string', () => {
    let input = 'a';
    let actual = getPermutations(input);
    let expected = new Set(['a']);
    expect(actual).to.deep.equal(expected);
  });

  it('two character string', () => {
    let input = 'ab';
    let actual = getPermutations(input);
    let expected = new Set(['ab', 'ba']);
    expect(actual).to.deep.equal(expected);
  });

  it('three character string', () => {
    let input = 'abc';
    let actual = getPermutations(input);
    let expected = new Set(['abc', 'acb', 'bac', 'bca', 'cab', 'cba']);
    expect(actual).to.deep.equal(expected);
  });
});

File: Perl/t/recursive_permutations.pl
use Test::More;

sub permutations {
    my $str = shift;
    if ( length($str) <= 1 ) { return [$str] }

    my @results = ();
    my @arr     = split //sm, $str;
    for ( my $i = 0 ; $i <= $#arr ; $i++ ) {
        my $char      = $arr[$i];
        my @remaining = ( @arr[ 0 .. $i - 1 ], @arr[ $i + 1 .. $#arr ] );

        my $perms = permutations( join '', @remaining );

        for my $p ( @{$perms} ) {
            push @results, $char . $p;
        }
    }

    return \@results;
}

sub assertPermutations {
    my ( $str, $expected, $message ) = @_;
    my $got = permutations($str);
    is_deeply( $got, $expected, $message );
}

assertPermutations( '',  [''],  'empty string' );
assertPermutations( 'a', ['a'], 'one character' );
assertPermutations( 'ab', [ 'ab', 'ba' ], 'two characters' );
assertPermutations(
    'abc',
    [ 'abc', 'acb', 'bac', 'bca', 'cab', 'cba' ],
    'three characters'
);

done_testing;
